app/src/main/java/com/mozilla/speechapp/MainActivity.java

```diff
@@ -75,7 +102,12 @@ private void initialize() {
                 mSeries1.resetData(new DataPoint[0]);
                 mMozillaSpeechService.setLanguage(txtLanguage.getText().toString());
                 mMozillaSpeechService.setProductTag(txtProdutTag.getText().toString());
-                mMozillaSpeechService.start(getApplicationContext());
+                mMozillaSpeechService.setModelPath(getExternalFilesDir("models").getAbsolutePath());
+                if (mMozillaSpeechService.ensureModelInstalled()) {
+                    mMozillaSpeechService.start(getApplicationContext());
+                } else {
+                    maybeDownloadOrExtractModel(getExternalFilesDir("models").getAbsolutePath(), mMozillaSpeechService.getLanguageDir());
```

Comment:
This logic is required on the app side because after checking with a few potential users they indeed want to be more in charge of securing / handling the selection and distribution of models.